### ISSUE TITLE
Lecture 3. Error analysis

### ISSUE THREAD

**Issue by seonghunn** [seonghunn/CS182#3]
# Part 1
### Empirical risk and true risk
**Risk : probability you will get it wrong** 즉, f<sub>&theta;</sub>(x)가 얼마나 오류가 있는지를 정량화 한 것. loss function의 평균값이다. 이는 training set의 loss와 같지 않은데,

<img width="435" alt="image" src="https://user-images.githubusercontent.com/69765278/159135384-a4b12af7-db4d-4ee4-aa1d-d2cfb0b0b371.png">

위에서의 설명과 같이, 우리는 확률분포 p(x) 를 모르며, 데이터 D만 알 뿐이다.
그래서 Empirical risk를 사용하는데, 이는 **true risk의 추정치** 라고 할 수 있다.
supervised learning은 이를 최소화하면서 학습하는 것 (empirical risk minimization)

true risk : test dataset 의 loss
empirical risk : train dataset 의 loss

### Overfitting
**empirical risk가 low하고, true risk가 high 한 것** 쉽게 생각하면, training set만 잘 예측하고, 이를 벗어난 데이터는 예측을 잘 못하는 모델이다.

<img width="180" alt="image" src="https://user-images.githubusercontent.com/69765278/159135525-3fef8027-f5a7-4fba-a6fc-181300b75809.png">

- dataset이 너무 작을 때 (일반화가 안 됨)
- 모델이 너무 강력할 때 (parameter가 너무 많아 주어진 training set에만 맞는 모델이 되어버림)

### Underfitting
**empirical risk가 high하고, true risk가 high 한 것**, 지나치게 일반화하여 잘 예측하지 못하는 모델이다.

<img width="93" alt="image" src="https://user-images.githubusercontent.com/69765278/159135594-ed94ff91-538c-4603-896e-697e81e637d4.png">

- model이 너무 약할 때 (parameter가 너무 작아서 예측을 잘 못함)
- optimizer가 맞지 않아서 학습을 잘 하지 못했을 때

이전 수업까지는 classfication을 다뤘는데, regression을 다뤄볼 것.
classification : object probability를 예측하는 것, 즉 정해진 클래스 (손글씨 숫자가 몇인지) 를 예측
regression : continuous distribution 을 예측하는 것. continuous한 distribution으로 매핑하는 것이다.

이 continuous distribution 중 가우스 분포에 대해 살펴보면, 가우스 분포는 평균과 분산의 함수이다. 이 때 평균은 대부분 '상수'가 아닌 '함수'가 된다.
식을 아래와 같이 전개하고, sigma를 I 로 고정한다면 다음과 같은 식 유도가 가능하다.

<img width="680" alt="image" src="https://user-images.githubusercontent.com/69765278/159135774-757795f3-c7b4-4360-bc8e-1d2c0426bc52.png">

이는 Mean Squared Error loss와 동일하다.

<img width="482" alt="image" src="https://user-images.githubusercontent.com/69765278/159135864-c5d8aec0-57a1-4580-8a51-22663c1146fe.png">

### How does the error change for different training sets?

다시 복기하면
true risk : test dataset 의 loss
empirical risk : train dataset 의 loss

<img width="630" alt="image" src="https://user-images.githubusercontent.com/69765278/159136083-bfd1c76c-8651-4b6f-8046-c65c7bbee499.png">

위 예시를 보면 점이 train dataset, 파랑 선이 test dataset이라 할 수 있다.
즉 overfitting 이란 train dataset은 전부 맞추어 empirical risk는 낮지만, 실제 test dataset인 파란 선과는 차이가 크기 때문에 true risk는 high 한 경우이며, 그렇기 때문에 다른 데이터를 넣으면 예측을 잘 하지 못할 것이다 (파란 선을 똑바로 예측한 것이 아니므로), 따라서 train dataset을 바꿔버리게 되면 모델이 변하게 된다.
차수가 필요 이상으로 높은 다항함수로 모델을 설계한 경우로 예시를 들 수 있다.
 underfitting는 둘 다 맞지 않는 경우이다. (두 risk가 high) 즉 모델의 parameter가 충분하지 않아 linear 하게 모델이 형성된 경우로, train data를 바꾸더라도 모델의 변화가 거의 없다 (예측을 잘 못하는 모델이기 떄문이다)

<img width="674" alt="image" src="https://user-images.githubusercontent.com/69765278/159136367-3340cfed-5dd3-42d0-b94f-b4ebd25b6ef1.png">

f<sub>&theta;</sub>(x) 는 의 theta 는 랜덤변수 D로부터 학습된 것이다.즉 theta는 각각의 D에 따라 달라지기 때문에 이를 f<sub>D</sub>(x)로 바꿀 수 있다. y는 notation의 통일을 위해 f(x)로 바꾸었는데 이는 해답에 해당하는 true function이다.
따라서 f(x)는 true function, f<sub>D</sub>(x)는 D라는 training dataset을 통해 학습한 함수이다.
기댓값은 확률과 확률변수의 곱이므로 아래와 같이 정리된다.

<img width="597" alt="image" src="https://user-images.githubusercontent.com/69765278/159136529-658b2981-7e0d-4030-a153-b3c111bd96c7.png">

위의 식에서 얻고자 하는 것은 우리의 학습 알고리즘이 dataset의 무작위한 선택에도 얼마나 독립적인지를 측정하기 위해서이다 (overfitting 처럼 dataset의 선택에 따라 모델이 자꾸 변하면 안됨)

위 식을 분석하기 위해, 아래와 같이 가정하자.

<img width="629" alt="image" src="https://user-images.githubusercontent.com/69765278/159136628-f5525234-7113-4db2-926d-90f86dbd34e1.png">

위의 식에서 f bar (x) 는 모든 possible dataset D에 대해서 결정되는 함수 f<sub>D</sub>(x)의 평균값을 의미한다.
식을 다음과 같이 조작할 수 있는데

<img width="602" alt="image" src="https://user-images.githubusercontent.com/69765278/159136774-6298d911-119d-450e-ab38-eb4fff2b458b.png">

f bar (x) 와 f(x)는 D에 영향을 받지 않으므로 평균을 계산할 때 상수처럼 빠져나올 수 있다. f bar (x)는 그 자체로 D에 대한 평균이기 때문에, D에 이미 영향을 받지 않는다고 할 수 있다. 따라서 f<sub>D</sub>(x) 의 평균은 f bar (x) 이므로, 다음과 같은 cancel out으로 항 하나가 삭제된다.

<img width="613" alt="image" src="https://user-images.githubusercontent.com/69765278/159136848-de3a45ac-a6ca-41dd-9055-37fd0babd2f6.png">

<img width="629" alt="image" src="https://user-images.githubusercontent.com/69765278/159136886-93de031b-39b1-4db0-86e1-e4bbabb42d7e.png">

따라서 위의 식을 얻게 되는데, 두번째 항의 f bar (x) 와 f(x)는 모두 D에 의존하지 않으므로, 평균이 단순한 값이 된다. 그렇게 정리한 위의 식에서 전자는 variance라 하고, 후자는 bias의 square 라고 한다.
bias는 true function에서 average가 얼마나 벗어나 있는지에 대한 양이다. 즉 예측값과 실제값이 얼마나 차이나는지를 측정한다. bias가 크면 underfitting이 일어난다.
variance는 true function이 무엇인지에 관계없이 dataset의 변화에 따라 예측 함수가 얼만큼 변하느냐에 대한 양이다. variance가 크면 overfitting이 일어난다.

이 둘은 trade off 관계이며, 이를 적절히 조절해야 한다.

# Part 2
## How to regulate bias/variance?
- 더 많은 데이터를 가져온다.
<img width="589" alt="image" src="https://user-images.githubusercontent.com/69765278/159137053-17e8def2-781e-4bdc-98ae-99b9e4db970d.png">
  variance에는 효과가 있지만, bias에는 효과가 없다.
- model class를 변화시킨다. (12차항 다항식을 선형식으로 바꾸는 등으로 단순화한다.
  이를 'Smooth' 하게 변화시키는 것이 **Regularization**이다.

## Regularization
**something we add to the loss function to reduce variance**

variance가 크다면, data가 parameter를 결정하는 데에 충분한 데이터를 주지 않았다고 해석할 수 있다.

<img width="469" alt="image" src="https://user-images.githubusercontent.com/69765278/159137461-4d6ed723-0603-44e4-b9c0-0842ac655101.png">

위의 세 가지 모델은 training error가 전부 zero다. 우리가 위의 세 가지 모델을 차별화 할 수 있는 정보를 준다면, 우리는 세가지 다 training error가 0이지만 가장 오른쪽 모델이 정답에 가장 가까운 모델이란 것을 알 수 있을 것이다.

여러 가지 interpretation 중, Bayesian intepretation 에 대해 살펴본다.
**Given D에 대해, 어떤 것이 가장 적절한 &theta; 인가?** 이를 p(&theta;|D)로 표현할 수 있다.

<img width="626" alt="image" src="https://user-images.githubusercontent.com/69765278/159137549-816bb5fc-5632-4e9b-8cfc-6d37aa3d702d.png">

prior는 p(&theta;)로, data D 입력 이전 &theta; 가 어땠는지에 대한 값이다.
우리는 이 prior를 함수가 smoother하게 만들기 위해서 다음과 같은 new loss function을 고려한다.

<img width="413" alt="image" src="https://user-images.githubusercontent.com/69765278/159137657-1e243a1e-4a97-4d81-877e-5022cfbc6707.png">

우리는 함수를 smooth 하게 만들기 위해서 다음과 같은 특성을 이용한다.

<img width="539" alt="image" src="https://user-images.githubusercontent.com/69765278/159137884-1239e33f-2fe6-49c9-9861-24251fbdd79a.png">

polynomial function에서, 각각의 계수가 크다면 왼쪽과 같은 함수 개형이, 각각의 계수가 작다면 오른쪽과 같은 함수 개형이 나온다. 여기서 prior로 Normal distribution 을 사용할건데 평균이 0이고, 최빈값이 0이기 때문에 계수를 작게 만들기에 굉장히 유용하기 때문이다. prior를 normal distribution으로 설정하면

<img width="305" alt="image" src="https://user-images.githubusercontent.com/69765278/159137941-e28bc654-8ea1-4ca8-b205-8754b2492d95.png">

와 같이 되고, 이를 error function에 대입하면 sigma 가 I라고 가정할 때

<img width="333" alt="image" src="https://user-images.githubusercontent.com/69765278/159137972-d6ca3f3b-a6eb-4323-8d58-1d5820d7b384.png">

위와 같이 정리할 수 있다. 이 때 앞의 항은 MSE 이고, 뒤의 항은 regulazation 항으로 파라미터의 norm square 를 작게 만들어 준다. 예시를 들어 보자.

<img width="596" alt="image" src="https://user-images.githubusercontent.com/69765278/159138078-8b61ccc6-4cbe-47fa-b7e2-e54da0cfc6d0.png">

위의 두 그림 모드 train dataset에 대하여 정답을 도출해 내고 있긴 하다. 하지만 왼쪽과 같은 모델을 원한다.
우리는 조금 더 smooth 하게 모델을 만들고 싶은데

<img width="642" alt="image" src="https://user-images.githubusercontent.com/69765278/159138125-8fb883b0-f2b3-4bc7-9749-545de5754a3f.png">

위와 같이 뾰족한 분포를 smooth하게 만들기 위해 작은 theta를 원한다. 따라서 p(&theta;) 로 normal distribution을 설정한다.
우리는 다양한 regularizers 를 이용하는데,

<img width="660" alt="image" src="https://user-images.githubusercontent.com/69765278/159138221-05757cc4-edf3-430d-be97-ef43a3183c66.png">

L1 regularization은 쓸모없는 계수를 **0** 으로 만드는 데에 효과적이다. 따라서 dimension을 큰 폭으로 줄일 수 있다. 이래서 다이아몬드 형태로 시각화 되는 것이다.
이외에도 Dropout, Gradient penalty 등등 여러 가지가 있다.

Bayesian 관점 이외에 Numerical 관점에서는 underdetermined 문제를 well determined 되도록 만든다고 이해할 수 도 있다. Optimization 은 loss landscap를 찾기 쉽게 만든다고 생각할 수 도 있다.

<img width="582" alt="image" src="https://user-images.githubusercontent.com/69765278/159138394-593ce899-f7dd-4f03-bbaf-323de09d5f42.png">

하이퍼파라미터들을 잘못 선택하면 최적해를 찾지 못할 수 있기 때문에, 이는 굉장히 중요하다.

# Part 3
overfitting , underfitting 극복, 알고리즘 선택, hyperparameter의 선택 등 여러가지 선택을 어떻게 해야할지에 대한 의문이 있다.
한 가지 아이디어는 loss 를 가장 작게 만드는 것인데 variance 와 bias 는 trade off 관계에 있기 때문에 이는 좋지 않은 선택이다. loss를 작게 만들어도 variance를 조절하지 못하기 때문에 overfitting을 잘 걸러내지 못할 것이다.
이를 위해 ML workflow를 확인해 보자.

<img width="609" alt="image" src="https://user-images.githubusercontent.com/69765278/159138509-11ffc81e-d47e-4df6-80fc-056c5ec57381.png">

80% 정도의 training set, 20% 정도의 validation set으로 구분할 수 있으며 training set을 통해 학습하고(모델 자체를 강화시킴) validation set을 통해 하이퍼파라미터 등을 결정하게 된다(학습전략을 결정한다).
training set을 통해 loss가 낮지 않다면 underfitting을 의심할 수 있고, regularization을 줄이거나 optimizer를 개선시키는 방법을 사용할 수 있다.
validation set을 사용했을 때의 loss가 train loss가 크다면 overfitting을 의심할 수 있다. 따라서 regularization을 증가시켜야 한다.

또한, training set을 통해 learning rate와 같은 optimizer hyperparameters 를 결정하며
validation set을 통해 model class 를 결정하며, regularizer hyperparameters 를 결정하게 된다.

<img width="532" alt="image" src="https://user-images.githubusercontent.com/69765278/159138736-05ba6dd0-7452-4ebb-857d-5e138d267382.png">

왼쪽 그림은 전형적인 overfitting 개형이다. (val loss 증가) 오른쪽 그림은 전형적인 underfitting 이다 (bias가 학습을 끝내도 줄어들지 않음)
우리가 왼쪽 그림에서 val loss가 증가하는 것을 보고 hyperparameter를 다시 설정하면 val set을 사용하여 개선시킨 것이기 때문에 우리가 성능을 테스트 할 때 그 val set을 사용할 수 없다.
이를 해결하기 위한 방법으로 **우리가 얼마나 model이 잘 설계되었는지 만을** 확인하기 위해 **test set**을 사용한다.

<img width="609" alt="image" src="https://user-images.githubusercontent.com/69765278/159138926-a3ba143a-1c0e-42b1-9d08-f8c288dffec4.png">

따라서 위와 같이 training set, validation set, test set 3 가지로 dataset을 구분하여 사용해야 한다.

## Summary

<img width="609" alt="image" src="https://user-images.githubusercontent.com/69765278/159138966-60990dda-f06c-471b-a5fa-fe8619f6a835.png">
